src/translate.py

```diff
@@ -431,29 +446,78 @@ def get_stack_info(
         ):
             # Saving the return address on the stack.
             info.is_leaf = False
-            info.return_addr_location = inst.args[1].lhs_as_literal()
+            stack_offset = inst.args[1].lhs_as_literal()
+            info.return_addr_location = stack_offset
+            callee_saved_words.append(stack_offset)
         elif (
             inst.mnemonic in ["sw", "swc1", "sdc1"]
             and destination in SAVED_REGS
             and isinstance(inst.args[1], AsmAddressMode)
             and inst.args[1].rhs.register_name == "sp"
+            and destination not in info.callee_save_reg_locations
         ):
             # Initial saving of callee-save register onto the stack.
-            info.callee_save_reg_locations[destination] = inst.args[1].lhs_as_literal()
-
-    # Find the region that contains local variables. It is above saved registers
-    # and the return address, if those exist. If they don't, the local variables
-    # can lie directly at the bottom of the stack.
-    info.local_vars_region_bottom = 0
+            stack_offset = inst.args[1].lhs_as_literal()
+            info.callee_save_reg_locations[destination] = stack_offset
+            if inst.mnemonic == "sdc1":
+                callee_saved_doubles.append(stack_offset)
+            else:
+                callee_saved_words.append(stack_offset)
+        elif (
+            inst.mnemonic in ["lw", "lwc1", "ldc1"]
+            and isinstance(inst.args[1], AsmAddressMode)
+            and inst.args[1].rhs.register_name == "sp"
+        ):
+            info.subroutine_arg_top = update_min(
+                info.subroutine_arg_top, inst.args[1].lhs_as_literal()
+            )
+        elif (
+            inst.mnemonic == "addiu"
+            and inst.args[1] == Register("sp")
+            and isinstance(inst.args[2], AsmLiteral)
+            and inst.args[2].value < info.allocated_stack_size
+        ):
+            info.subroutine_arg_top = update_min(
+                info.subroutine_arg_top, inst.args[2].value
+            )
 
     if not info.is_leaf:
-        info.local_vars_region_bottom = info.return_addr_location + 4
+        # Compute the bounds of the callee-saved register region, including padding
+        callee_saved_words.sort()
+        callee_saved_doubles.sort()
+        bottom = callee_saved_words[0]
+
+        # Both GCC & IDO save registers in two blocks: one for word-sized registers,
+        # followed by one for doubles. The word-sized block is also padded to a
+        # multiple of 8 bytes. Check that there are no gaps in either of these blocks.
+        top = bottom
+        for offset in callee_saved_words:
+            assert offset == top, DecompFailure(
```

Comment:
this still raises an AssertionError, change to `if offset != top: raise DecompFailure(...)`